### PR TITLE
Add refactor of filesystem code

### PR DESCRIPTION
**Pull request by fearful-symmetry**
This is basically a copy-and-paste of https://github.com/elastic/beats/pull/31001 (or at least the portions that should go here).

This is part of a the ongoing refactor of the system code, and it includes all the metrics needed to support the system/filesystem and system/fsstat metricsets.

### CODE REVIEW

**Comment by cmacknz** [comment]
Changes look good (really I believe that you copied them :) )

Let's add a license file in this commit since the repo needs one. You can just copy the Apache license file from beats: https://github.com/elastic/beats/blob/main/LICENSE.txt

**Comment by fearful-symmetry** [comment]
Okay, so, I forgot that PR also added some code to stuff that's now in `elastic-agent-libs`, so this CI is gonna fail until I put in a PR there.

File: LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Source code in this repository is variously licensed under the Apache License
```
Comment (cmacknz) [inline LICENSE.txt:1]: Ugh, I forgot that beats used a dual license. We want the plain Apache license: https://github.com/elastic/elastic-agent-libs/blob/main/LICENSE

Sorry for the wrong suggestion.

File: metric/system/filesystem/filesystem_test.go [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+	"os"
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/elastic-agent-libs/mapstr"
+
+	"github.com/elastic/elastic-agent-libs/transform/typeconv"
+	"github.com/elastic/elastic-agent-system-metrics/metric/system/resolve"
```
Comment (belimawr) [inline metric/system/filesystem/filesystem_test.go:31]: I'm not sure if there is a lint already setup to this repo, but those lines could be a single block
```suggestion
	"github.com/elastic/elastic-agent-libs/mapstr"
	"github.com/elastic/elastic-agent-libs/transform/typeconv"
	"github.com/elastic/elastic-agent-system-metrics/metric/system/resolve"
```

File: metric/system/filesystem/filesystem_test.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+	}
+	assert.True(t, (len(fss) > 0))
+
+	for _, fs := range fss {
+
+		err := fs.GetUsage()
+
+		assert.NoError(t, err, "filesystem=%v: %v", fs, err)
+
+	}
```
Comment (belimawr) [inline metric/system/filesystem/filesystem_test.go:70]: Small style/lint:
```suggestion
	for _, fs := range fss {
		err := fs.GetUsage()
		assert.NoError(t, err, "filesystem=%v: %v", fs, err)
	}
```

File: metric/system/filesystem/filesystem_windows.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package filesystem
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/v7/libbeat/opt"
```
Comment (belimawr) [inline metric/system/filesystem/filesystem_windows.go:23]: Should this be `"github.com/elastic/elastic-agent-libs/opt"`?
```suggestion
	"github.com/elastic/elastic-agent-libs/opt"
```

**Review by cmacknz — APPROVED** [review]

**Pull request merged**